test/Core.Test/Services/LicensingServiceTests.cs

```diff
@@ -38,12 +51,215 @@ public LicensingServiceTests()
+            {
+                Id = organizationId,
+                Enabled = true,
+                LicenseKey = "test",
+                PlanType = PlanType.Custom,
+                Seats = 2,
+                MaxCollections = 1,
+                UseGroups = true,
+                UseDirectory = true,
+                UseTotp = true,
+                SelfHost = true,
+                Name = "Test"
+            };
+            
+            var organizations = new List<Organization> { organization };
+            _organizationRepository.GetManyByEnabledAsync().Returns(organizations);
+
+            Directory.CreateDirectory(_organizationDirectory);
+            var license = Substitute.For<OrganizationLicense>();
+            license.LicenseKey = "test";
+            license.Version = 1;
+            license.Enabled = organization.Enabled;
+            license.Issued = DateTime.UtcNow.AddDays(-1);
+            license.Expires = DateTime.UtcNow.AddDays(1);
+            license.InstallationId = _globalSettings.Installation.Id;
+            license.PlanType = organization.PlanType;
+            license.Seats = organization.Seats;
+            license.MaxCollections = organization.MaxCollections;
+            license.UseGroups = organization.UseGroups;
+            license.UseDirectory = organization.UseDirectory;
+            license.UseTotp = organization.UseTotp;
+            license.SelfHost = organization.SelfHost;
+            license.Name = organization.Name;
+            license.Signature = string.Empty;
+            
+            var licenseFile = CreateLicenseFile(organizationId, license);
+            
+            // Act
+            await _sut.ValidateOrganizationsAsync();
+            
+            // Assert
+            await _organizationRepository.Received().ReplaceAsync(organizations.First());
+            await _organizationRepository.Received().ReplaceAsync(Arg.Is<Organization>(o => o.Enabled == false));
+
+            
+            var loggerArguments = _logger.ReceivedCalls().Last().GetArguments()
+                                                    .Select(arg => arg.ToString())
+                                                    .Where(arg => arg != null)
+                                                    .AsEnumerable<string>();
+            
+            Assert.Contains(loggerArguments, arg => arg.Contains("Invalid signature"));
+            
+            // Tear down
+            DeleteLicenseFile(licenseFile);
+        }
+        
+        private static string CreateLicenseFile(Guid organizationId, OrganizationLicense license)
+        {
+            var fileName = $"{_organizationDirectory}/{organizationId}.json";
+
+            Directory.CreateDirectory(_organizationDirectory);
+
+            File.WriteAllText(fileName, JsonConvert.SerializeObject(license));
+
+            return fileName;
+        }
+
+        private static void DeleteLicenseFile(string licenseFile)
```

Comment:
I would be curious to know if there's a better way to mock or fake file system access for these license files through the license service vs. writing/reading physical files for unit tests.

Also, I don't see test hooks for overall test execution cleanup to check for errant files and delete them.

Reply:
That's fair. I looked into loading in an In-Memory file system, but the ones I found were relying on being able to swap out the FileSystem using System.IO.Abstractions. I didn't really want to refactor the existing code without tests, so this seemed like a good stepping point.

I'll work on a safe way of cleaning up the directory without affecting the other tests that will run at the same time.